### ISSUE TITLE
Feature: Arbitrary range types

### ISSUE THREAD

**Issue by ElectronicRU** [epgsql/epgsql#122]
PostgreSQL has multiple built-in range types, and allows creating a new one over just about anything comparable. Currently only a couple of range types are supported, and as a special case.

I propose the generic support for range types be added, as {range, whatever}, since the encoding for them is rather type-agnostic, and the corresponding information can be optained from pg_range system table.

Please provide feedback on whether you'd think this feature would be useful.

**Comment by seriyps** [comment]
Yep, sounds good. At least elixir driver have generic range implementation https://github.com/elixir-ecto/postgrex/blob/master/lib/postgrex/extensions/range.ex

I think it may be implemented in the way how arrays are implemented (some time ago epgsql had all this `int8array`, `timestamparray` etc datatypes) https://github.com/epgsql/epgsql/pull/4

**Comment by ElectronicRU** [comment]
Unlike Postgrex, epgsql currently uses type name and type structure interchangeably. This means that a named range type would either be not usable as named, or as a range type.
Some mapping from name to structure might be in order.

**Comment by ElectronicRU** [comment]
Empty ranges don't seem to be handled right now, too, and positive_infinity and negative_infinity seem to be unique atoms that go against the convention (could just use infinity unambiguously). Is it too late to rethink the representation?

**Comment by seriyps** [comment]
I think we may introduce some connection option like `{type_opts, [{range, ...}, {timestamp, ...}]}` to change type encoding/decoding behaviour.
We have also #109, which may solve this. But I don't know when exactly I or somebody else will have time to implement this.

**Comment by ElectronicRU** [comment]
Concerning #109, I might be willing to take the coding of it if I get help/review with the design document beforehand.

**Comment by seriyps** [comment]
@ElectronicRU would be nice. Do you think we should use github's Wiki page for design doc?

**Comment by ElectronicRU** [comment]
Well, they're versioned, and Markdown isn't horrible (who am i kidding, it is), so that's an option, yes.

**Comment by tsloughter** [comment]
Noticed this issue, it would be resolved with #180 :)
